File: src/main/java/com/customagent/service/TestGenerationService.java
package com.customagent.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * Pure-logic service that assists in generating JUnit 5 test templates
 * and analysing source code for test coverage gaps.
 * <p>
 * This service has <em>no</em> external dependencies and is therefore
 * straightforward to unit-test without mocking.
 */
@Slf4j
@Service
public class TestGenerationService {

    private static final Set<String> TEST_ANNOTATION_KEYWORDS = Set.of(
            "@Test", "@ParameterizedTest", "@RepeatedTest"
    );

    // ─── TEMPLATE GENERATION ─────────────────────────────────────────────────

    /**
     * Generates a JUnit 5 test-class skeleton for the given class and its
     * public method names.
     *
     * @param className   simple class name (e.g. "UserService")
     * @param methodNames non-null, non-empty list of public method names
     * @return complete test-class template as a String
     * @throws IllegalArgumentException if className or methodNames is null/empty
     */
    public String generateTestTemplate(String className, List<String> methodNames) {
        if (className == null || className.isBlank()) {
            throw new IllegalArgumentException("Class name cannot be null or blank");
        }
        if (methodNames == null || methodNames.isEmpty()) {
            throw new IllegalArgumentException(
                    "Method names list cannot be null or empty");
        }

        StringBuilder sb = new StringBuilder();
        sb.append("@ExtendWith(MockitoExtension.class)\n");
        sb.append("class ").append(className).append("Test {\n\n");
        sb.append("    @InjectMocks\n");
        sb.append("    private ").append(className).append(" subject;\n\n");

        for (String method : methodNames) {
            sb.append("    @Test\n");
            sb.append("    void ").append(method).append("_shouldSucceed() {\n");
            sb.append("        // Arrange\n");
            sb.append("        // Act\n");
            sb.append("        // Assert\n");
            sb.append("    }\n\n");
        }

        sb.append("}");
        log.info("Generated test template for class={} methodCount={}",
                className, methodNames.size());
        return sb.toString();
    }

    // ─── COVERAGE ANALYSIS ───────────────────────────────────────────────────

    /**
     * Scans source-code lines and returns names of {@code public} methods that
     * are <em>not</em> immediately preceded by a recognised test annotation.
     *
     * @param sourceCode raw Java source code string (may be null)
     * @return unmodifiable list of uncovered method names; never null
     */
    public List<String> identifyUncoveredMethods(String sourceCode) {
        if (sourceCode == null || sourceCode.isBlank()) {
            return Collections.emptyList();
        }

        String[] lines = sourceCode.split("\n");
        List<String> uncovered = new ArrayList<>();

        for (int i = 0; i < lines.length; i++) {
            String trimmed = lines[i].trim();
            if (isPublicMethodDeclaration(trimmed)) {
                boolean hasAnnotation = (i > 0)
                        && hasTestAnnotation(lines[i - 1].trim());
                if (!hasAnnotation) {
                    String name = extractMethodName(trimmed);
                    if (name != null) {
                        uncovered.add(name);
                    }
                }
            }
        }
        return Collections.unmodifiableList(uncovered);
    }

    // ─── STATS ───────────────────────────────────────────────────────────────

    /**
     * Classifies a list of test method names into happy-path vs error scenarios
     * based on name keywords (error, exception, invalid, null, fail, negative).
     *
     * @param testMethodNames non-null list of test method name strings
     * @return unmodifiable map with keys: {@code total}, {@code happyPath},
     *         {@code errorScenarios}
     * @throws IllegalArgumentException if testMethodNames is null
     */
    public Map<String, Integer> calculateTestStats(List<String> testMethodNames) {
        if (testMethodNames == null) {
            throw new IllegalArgumentException("Test methods list cannot be null");
        }

        long errorCount = testMethodNames.stream()
                .filter(m -> m.toLowerCase(Locale.ROOT)
                        .matches(".*(error|exception|invalid|null|fail|negative).*"))
                .count();

        Map<String, Integer> stats = new LinkedHashMap<>();
        stats.put("total",          testMethodNames.size());
        stats.put("errorScenarios", (int) errorCount);
        stats.put("happyPath",      testMethodNames.size() - (int) errorCount);
        return Collections.unmodifiableMap(stats);
    }

    // ─── PRIVATE HELPERS ─────────────────────────────────────────────────────

    private boolean isPublicMethodDeclaration(String line) {
        return line.contains("public ")
                && line.contains("(")
                && !line.startsWith("//")
                && !line.startsWith("*")
                && !line.startsWith("/*")
                && !line.contains("class ")
                && !line.contains("interface ");
    }

    private boolean hasTestAnnotation(String line) {
        return TEST_ANNOTATION_KEYWORDS.stream().anyMatch(line::startsWith);
    }

    private String extractMethodName(String line) {
        try {
            int paren = line.indexOf('(');
            if (paren <= 0) return null;
            String[] parts = line.substring(0, paren).trim().split("\\s+");
            return parts[parts.length - 1];
        } catch (Exception e) {
            return null;
        }
    }
}

File: src/services/testGenerationService.js
'use strict';

/**
 * TestGenerationService — pure-logic service for generating JUnit 5 test
 * templates and analysing source code for coverage gaps.
 *
 * No external dependencies; mirrors the Java TestGenerationService exactly.
 */

const TEST_ANNOTATION_KEYWORDS = new Set(['@Test', '@ParameterizedTest', '@RepeatedTest']);

// ── TEMPLATE GENERATION ─────────────────────────────────────────────────────

/**
 * Generates a JUnit 5 test-class skeleton for the given class and its
 * public method names.
 *
 * @param {string}   className   Simple class name (e.g. "UserService")
 * @param {string[]} methodNames Non-null, non-empty list of public method names
 * @returns {string} Complete test-class template
 * @throws {Error} If className is null/blank or methodNames is null/empty
 */
function generateTestTemplate(className, methodNames) {
  if (!className || className.trim() === '') {
    throw new Error('Class name cannot be null or blank');
  }
  if (!methodNames || !Array.isArray(methodNames) || methodNames.length === 0) {
    throw new Error('Method names list cannot be null or empty');
  }

  const lines = [];
  lines.push('@ExtendWith(MockitoExtension.class)');
  lines.push(`class ${className}Test {`);
  lines.push('');
  lines.push('    @InjectMocks');
  lines.push(`    private ${className} subject;`);
  lines.push('');

  for (const method of methodNames) {
    lines.push('    @Test');
    lines.push(`    void ${method}_shouldSucceed() {`);
    lines.push('        // Arrange');
    lines.push('        // Act');
    lines.push('        // Assert');
    lines.push('    }');
    lines.push('');
  }

  lines.push('}');
  return lines.join('\n');
}

// ── COVERAGE ANALYSIS ────────────────────────────────────────────────────────

/**
 * Scans source-code lines and returns names of `public` methods that are NOT
 * immediately preceded by a recognised test annotation.
 *
 * @param {string|null} sourceCode Raw Java source code string (may be null)
 * @returns {string[]} Unmodifiable list of uncovered method names
 */
function identifyUncoveredMethods(sourceCode) {
  if (!sourceCode || sourceCode.trim() === '') {
    return [];
  }

  const lines = sourceCode.split('\n');
  const uncovered = [];

  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].trim();
    if (isPublicMethodDeclaration(trimmed)) {
      const prevLine = i > 0 ? lines[i - 1].trim() : '';
      const hasAnnotation = hasTestAnnotation(prevLine);
      if (!hasAnnotation) {
        const name = extractMethodName(trimmed);
        if (name !== null) {
          uncovered.push(name);
        }
      }
    }
  }

  return uncovered;
}

// ── STATS ────────────────────────────────────────────────────────────────────

/**
 * Classifies test method names into happy-path vs error scenarios.
 *
 * @param {string[]} testMethodNames Non-null list of test method name strings
 * @returns {{ total: number, errorScenarios: number, happyPath: number }}
 * @throws {Error} If testMethodNames is null
 */
function calculateTestStats(testMethodNames) {
  if (testMethodNames === null || testMethodNames === undefined) {
    throw new Error('Test methods list cannot be null');
  }

  const errorKeywords = /error|exception|invalid|null|fail|negative/i;
  const errorCount = testMethodNames.filter((m) => errorKeywords.test(m)).length;

  return Object.freeze({
    total: testMethodNames.length,
    errorScenarios: errorCount,
    happyPath: testMethodNames.length - errorCount,
  });
}

// ── PRIVATE HELPERS ─────────────────────────────────────────────────────────

function isPublicMethodDeclaration(line) {
  return (
    line.includes('public ') &&
    line.includes('(') &&
    !line.startsWith('//') &&
    !line.startsWith('*') &&
    !line.startsWith('/*') &&
    !line.includes('class ') &&
    !line.includes('interface ')
  );
}

function hasTestAnnotation(line) {
  for (const kw of TEST_ANNOTATION_KEYWORDS) {
    if (line.startsWith(kw)) return true;
  }
  return false;
}

function extractMethodName(line) {
  try {
    const paren = line.indexOf('(');
    if (paren <= 0) return null;
    const parts = line.substring(0, paren).trim().split(/\s+/);
    return parts[parts.length - 1] || null;
  } catch (_) {
    return null;
  }
}

module.exports = {
  generateTestTemplate,
  identifyUncoveredMethods,
  calculateTestStats,
};
